run/storage-streaming/chunked-http.js

```diff
@@ -0,0 +1,111 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+// [START functions_tips_chunked_download]
+// [START cloudrun_tips_chunked_download]
+// [START functions_tips_chunked_upload]
+// [START cloudrun_tips_chunked_upload]
+const uuidv4 = require('uuid').v4;
+const Busboy = require('busboy');
+
+const {Storage} = require('@google-cloud/storage');
+const storage = new Storage();
+
+const sourceFile = storage.bucket('cloud-devrel-public').file('data.txt');
+
+// [END functions_tips_chunked_download]
+// [END cloudrun_tips_chunked_download]
+// TODO(developer): set this to a bucket you own
+const TARGET_BUCKET = process.env.FUNCTIONS_BUCKET;
+// [END functions_tips_chunked_upload]
+// [END cloudrun_tips_chunked_upload]
+// [START functions_tips_chunked_download]
+// [START cloudrun_tips_chunked_download]
+
+// This function downloads a file from Google Cloud Storage with HTTP chunking.
+// This lets a function send files that are bigger than the instance's available
+// memory.
+exports.chunkedDownload = async (req, res) => {
+  const readStream = sourceFile.createReadStream();
+
+  // Download file to the end-user's device
+  // TODO(developer): remove this to send files (e.g. images) directly to users
+  res.attachment('data.txt');
+
+  readStream.pipe(res);
+
+  return new Promise((resolve, reject) => {
+    readStream.on('error', reject).on('finish', resolve);
+  });
+};
+
+// [END functions_tips_chunked_download]
+// [END cloudrun_tips_chunked_download]
+
+// [START functions_tips_chunked_upload]
+// [START cloudrun_tips_chunked_upload]
+
+// This function uploads an HTTP-chunked file to Google Cloud Storage.
+// This lets you process files that are bigger than the instance's available
+// memory.
+exports.chunkedUpload = async (req, res) => {
+  if (req.method !== 'POST') {
+    // Return a "method not allowed" error
+    return res.status(405).end();
+  }
+
+  // Add an ID to the uploaded filename to prevent filename collisions.
+  const fileId = req.query.fileId || uuidv4();
+
+  const busboy = new Busboy({headers: req.headers});
+
+  const ioPromises = [];
+
+  busboy.on('file', (_, localFile, fileName) => {
+    const idName = `chunked-http-${fileId}-${fileName}`;
+
+    const targetFile = storage.bucket(TARGET_BUCKET).file(idName);
+    const writeStream = targetFile.createWriteStream();
+
+    localFile.pipe(writeStream);
+
+    // File was processed by Busboy; wait for Cloud Storage upload to finish.
+    ioPromises.push(
+      new Promise((resolve, reject) => {
+        localFile.on('end', () => {
+          writeStream.end();
+        });
+        writeStream.on('finish', resolve).on('error', reject);
+      })
+    );
+  });
+
+  busboy.end(req.rawBody);
```

Comment:
what does .end do?

Reply:
Honestly, I don't know. Maybe @fhinkel or @bcoe does?

Reply:
busboy inherits from a WriteableStream, so `busboy.end(req.rawBody)` writes the raw bytes to the stream, and indicates that no more bytes will be written.

I feel like `req.pipe(busboy)` might be more idiomatic, and you wouldn't need to handle the `.end`:

Reply:
I tried that, and the tests failed (the GCS upload didn't go through).